hydrolib/core/io/ini/serializer.py

```diff
@@ -0,0 +1,245 @@
+        self, header_blocks: Iterable[CommentBlock]
+    ) -> Iterable[str]:
+        blocks_as_lines = (
+            self._serialize_comment_block(block) for block in header_blocks
+        )
+        # We separate comment blocks in the header with an empty line.
+        blocks_as_lines = Serializer._interweave(blocks_as_lines, [""])  # type: ignore
+
+        return chain.from_iterable(blocks_as_lines)
+
+    def _serialize_section_header(self, section_header: str) -> Iterable[str]:
+        indent = " " * (self._config.section_indent)
+        yield f"{indent}[{section_header}]"
+
+    @staticmethod
+    def _get_offset(key: Optional[str], max_length: int) -> str:
+        key_length = len(key) if key is not None else 0
+        return " " * max(max_length - key_length, 0)
+
+    def _serialize_property(
+        self, property: Property, lengths: _Lengths
+    ) -> Iterable[str]:
+        indent = " " * (self._config.total_property_indent)
+        key_offset = Serializer._get_offset(property.key, lengths.max_key_length)
+        key = f"{property.key}{key_offset} = "
+
+        value_offset = Serializer._get_offset(property.value, lengths.max_value_length)
+        value = (
+            f"{property.value}{value_offset}"
+            if property.value is not None
+            else value_offset
+        )
+
+        comment = f" # {property.comment}" if property.comment is not None else ""
+
+        yield f"{indent}{key}{value}{comment}".rstrip()
+
+    def _serialize_content(
+        self, content: Iterable[Union[Property, CommentBlock]], lengths: _Lengths
+    ) -> Iterable[str]:
+        def serialize_element(elem: Union[Property, CommentBlock]) -> Iterable[str]:
+            if isinstance(elem, Property):
+                return self._serialize_property(elem, lengths)
+            else:
+                return self._serialize_comment_block(
+                    elem, self._config.total_property_indent
+                )
+
+        return chain.from_iterable((serialize_element(elem) for elem in content))
+
+    def _serialize_datablock(
+        self, datablock: Optional[Iterable[Sequence[str]]], lengths: _Lengths
+    ) -> Iterable[str]:
+        if datablock is None or lengths.max_datablock_element_length is None:
+            return []
+
+        def serialize_row_element(elem: str, index: int) -> str:
+            offset = Serializer._get_offset(elem, lengths.max_datablock_element_length[index])  # type: ignore
+            return elem + offset
+
+        indent = " " * self._config.total_datablock_indent
+
+        def serialize_row(row: Sequence[str]) -> str:
+            return (
+                indent
+                + (" " * self._config.datablock_spacing)
+                .join((serialize_row_element(elem, i)) for elem, i in zip(row, count()))
+                .rstrip()
+            )
+
+        return (serialize_row(row) for row in datablock)
+
+    def _serialize_section(self, section: Section) -> Iterable[str]:
```

Comment:
Not specific to this section, but it would be good to have a call stack diagram for these parsers and serializers. There can be many private methods that are hard to understand because they don't have a context, i.e. where in the process are they used, what assumptions are made etc.

Reply:
Maybe something to add to #40 as well.